packages/remote/src/electron-node/setup/remote-setup-service.ts

```diff
@@ -0,0 +1,163 @@
+            await this.unzipRemote(connection, remoteNodeZip, applicationDirectory);
+        }
+        // 4. Copy backend to remote system
+        const libDir = RemotePlatform.joinPath(platform, applicationDirectory, 'lib');
+        const libDirExists = await this.dirExistsRemote(connection, libDir);
+        if (!libDirExists) {
+            report('Installing application on remote...');
+            const applicationZipFile = RemotePlatform.joinPath(platform, applicationDirectory, `${this.getRemoteAppName()}.tar`);
+            await this.copyService.copyToRemote(connection, platform, applicationZipFile);
+            await this.unzipRemote(connection, applicationZipFile, applicationDirectory);
+        }
+        // 5. start remote backend
+        report('Starting application on remote...');
+        const port = await this.startApplication(connection, platform, applicationDirectory, remoteNodeDirectory);
+        connection.remotePort = port;
+    }
+
+    protected async startApplication(connection: RemoteConnection, platform: RemotePlatform, remotePath: string, nodeDir: string): Promise<number> {
+        const nodeExecutable = RemotePlatform.joinPath(platform, nodeDir, 'bin', platform === 'windows' ? 'node.exe' : 'node');
+        const mainJsFile = RemotePlatform.joinPath(platform, remotePath, 'lib', 'backend', 'main.js');
+        const localAddressRegex = /listening on http:\/\/127.0.0.1:(\d+)/;
+        // Change to the remote application path and start a node process with the copied main.js file
+        // This way, our current working directory is set as expected
+        const result = await connection.execPartial(`cd "${remotePath}";${nodeExecutable}`,
+            stdout => localAddressRegex.test(stdout),
+            [mainJsFile, '--hostname=127.0.0.1', '--port=0', '--remote']);
+
+        const match = localAddressRegex.exec(result.stdout);
+        if (!match) {
+            throw new Error('Could not start remote system: ' + result.stdout);
+        } else {
+            return Number(match[1]);
+        }
+    }
+
+    protected async detectRemotePlatform(connection: RemoteConnection): Promise<RemotePlatform> {
+        const result = await connection.exec('uname -s');
+
+        if (result.stderr) {
+            // Only Windows systems return an error output here
+            return 'windows';
+        } else if (result.stdout) {
+            if (result.stdout.includes('windows32') || result.stdout.includes('MINGW64')) {
+                return 'windows';
+            } else if (result.stdout.includes('Linux')) {
+                return 'linux';
+            } else if (result.stdout.includes('Darwin')) {
+                return 'darwin';
+            }
+        }
+        throw new Error('Failed to identify remote system: ' + result.stdout + '\n' + result.stderr);
+    }
+
+    protected async getRemoteHomeDirectory(connection: RemoteConnection, platform: RemotePlatform): Promise<string> {
```

Comment:
All of these methods dispatch on the platform. Why not abstract the differences in a platform-specific "scripts" objects that we instantiate when we detect the platform? (StrategyObject pattern)